Encoder/DefaultEncoder.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+namespace Lexik\Bundle\JWTAuthenticationBundle\Encoder;
+
+use InvalidArgumentException;
+use Lexik\Bundle\JWTAuthenticationBundle\Services\KeyLoader\KeyLoaderInterface;
+use Namshi\JOSE\JWS;
+
+/**
+ * Default class for encode/decode JWT, based on the Namshi/JOSE library.
+ * Supports OpenSSL and PHPSecLib encryption engines.
+ *
+ * @since 2.0
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ */
+class DefaultEncoder extends JWTEncoder
+{
+    /**
+     * @var KeyLoaderInterface
+     */
+    protected $keyLoader;
+
+    /**
+     * @var string
+     */
+    protected $encryptionEngine;
+
+    /**
+     * @var string
+     */
+    protected $encryptionAlgorithm;
+
+    /**
+     * Constructor.
+     *
+     * @param KeyLoaderInterface $keyLoader
+     * @param string             $encryptionEngine
+     * @param string             $encryptionAlgorithm
+     *
+     * @throws InvalidArgumentException If the given encryptionEngine is not supported.
+     * @throws InvalidArgumentException If the given algorithm is not supported.
+     */
+    public function __construct(KeyLoaderInterface $keyLoader, $encryptionEngine, $encryptionAlgorithm)
+    {
+        if (!in_array($encryptionEngine, ['OpenSSL', 'SecLib'])) {
+            throw new InvalidArgumentException(
+                sprintf('The encryption engine "%s" is not supported', $encryptionEngine)
+            );
+        }
+
+        if (!$this->isAlgorithmSupportedForEngine($encryptionEngine, $encryptionAlgorithm)) {
+            throw new InvalidArgumentException(
+                sprintf('The algorithm "%s" is not supported for %s', $encryptionAlgorithm, $encryptionEngine)
+            );
+        }
+
+        $this->keyLoader           = $keyLoader;
+        $this->encryptionEngine    = $encryptionEngine;
+        $this->encryptionAlgorithm = $encryptionAlgorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \RuntimeException If the JWS cannot be signed
+     */
+    public function encode(array $data)
+    {
+        $jws = new JWS(['alg' => $this->encryptionAlgorithm], $this->encryptionEngine);
+        $key = $this->keyLoader->loadKey('private');
+
+        $jws->setPayload($data);
+
+        try {
+            $jws->sign($key, $this->keyLoader->getPassphrase());
+        } catch (InvalidArgumentException $e) {
+            throw new \RuntimeException(
```

Comment:
I tried to test most of possible use cases, the exception thrown by namshi/jose say "Invalid key supplied" if it can't use the given key with its passphrase, so I added this try-catch to give an indication about what is wrong and from where the problem come from, I also give the message of the previous exception in the message of the last thrown exception.
Do you think it makes sense to have this catch?